impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/InternalImpactedTestEngine.java

```diff
@@ -0,0 +1,111 @@
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.ServiceLoader;
+
+import static com.teamscale.test_impacted.engine.ImpactedTestEngine.ENGINE_ID;
+
+/**
+ * Test engine called internally to allow testing without needing a {@link ServiceLoader} for {@link TestEngine} setup.
+ */
+class InternalImpactedTestEngine implements TestEngine {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(InternalImpactedTestEngine.class);
+
+	private final TestEngineRegistry testEngineRegistry;
+
+	private final ITestExecutor testExecutor;
+
+	private final TestDataWriter testDataWriter;
+
+	InternalImpactedTestEngine(TestEngineRegistry testEngineRegistry,
+							   ITestExecutor testExecutor,
+							   TestDataWriter testDataWriter) {
+		this.testEngineRegistry = testEngineRegistry;
+		this.testExecutor = testExecutor;
+		this.testDataWriter = testDataWriter;
+	}
+
+	@Override
+	public String getId() {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
+		EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "Teamscale Impacted Tests");
+
+		LOGGER.debug(() -> "Starting test discovery for engine " + ENGINE_ID);
+
+		for (TestEngine delegateTestEngine : testEngineRegistry) {
+			LOGGER.debug(() -> "Starting test discovery for delegate engine: " + delegateTestEngine.getId());
+			TestDescriptor delegateEngineDescriptor = delegateTestEngine.discover(discoveryRequest,
+					UniqueId.forEngine(delegateTestEngine.getId()));
+
+			engineDescriptor.addChild(delegateEngineDescriptor);
+		}
+
+		LOGGER.debug(() -> "Discovered test descriptor for engine " + ENGINE_ID + ":\n" + TestDescriptorUtils
+				.getTestDescriptorAsString(engineDescriptor));
+
+		return engineDescriptor;
+	}
+
+	@Override
+	public void execute(ExecutionRequest request) {
+		EngineExecutionListener engineExecutionListener = request.getEngineExecutionListener();
+		EngineDescriptor engineDescriptor = (EngineDescriptor) request.getRootTestDescriptor();
+
+		LOGGER.debug(() -> "Starting execution of request for engine " + ENGINE_ID + ":\n" + TestDescriptorUtils
+				.getTestDescriptorAsString(engineDescriptor));
+
+		engineExecutionListener.executionStarted(engineDescriptor);
+		runTestExecutor(request);
+		engineExecutionListener.executionFinished(engineDescriptor, TestExecutionResult.successful());
+	}
+
+	private void runTestExecutor(ExecutionRequest request) {
+		List<TestDetails> availableTests = new ArrayList<>();
+		List<TestExecution> testExecutions = new ArrayList<>();
+
+		for (TestDescriptor engineTestDescriptor : request.getRootTestDescriptor().getChildren()) {
+			Optional<String> engineId = engineTestDescriptor.getUniqueId().getEngineId();
+
+			if (!engineId.isPresent()) {
+				continue;
```

Comment:
Does this indicate some kind of error state/bug or is this expected. If possible please add a comment on when this would happen.